### Model Description:
- LSA + TF-IDF (LSA is is SVD on doc-term matrix)
- Create function to create batches of doc intrusion
- This version removes duplicate tweets
- This version randomizes the order of the intruders in the output

In [1]:
import pandas as pd
import nltk
from nltk.tokenize import sent_tokenize, word_tokenize, RegexpTokenizer
from nltk.util import ngrams
nltk.download('stopwords')
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD
import re
import numpy as np
import random

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/jennihawk/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [2]:
#got tweets at 1:15 Pacific Time 10/14/22
tweets = pd.read_csv('/Users/jennihawk/Documents/Data Science Projects/Project_NLP/TweetBatch3.csv')
tweets

text  \
0      @ReallyAmerican1 #Roevember and\n#ForThePeople...   
1      RT @sandibachom: IS THIS THING ON???!!This is ...   
2      RT @sandibachom: IS THIS THING ON???!!This is ...   
3      RT @tleehumphrey: Today is the beginning of th...   
4      RT @AdamKinzinger: Mitch McConnell.\nKevin McC...   
...                                                  ...   
34988  RT @Adrian_Fontes: The January 6th committee j...   
34989  #January6thCommitteeHearings and everyone runn...   
34990  RT @sandibachom: IS THIS THING ON???!!This is ...   
34991  So they are gonna subpoena Trump I am guessing...   
34992  RT @mariashriver: This is a historic day. @Rep...   

                                                 cleaned  
0       roevember and forthepeople and votebluein2022...  
1      rt is this thing on this is pathetic acting se...  
2      rt is this thing on this is pathetic acting se...  
3      rt today is the beginning of the inquiry into ...  
4      rt mitch mcconnell kevin mccarthy they both kn...  
...                                                  ...  
34988  rt fontes the january 6th committee just concl...  
34989   january6thcommitteehearings and everyone runn...  
34990  rt is this thing on this is pathetic acting se...  
34991  so they are gonna subpoena trump am guessing t...  
34992  rt this is historic day words were somber rega...  

[34993 rows x 2 columns]

### Remove stop words from cleaned column

In [3]:
stop_words = stopwords.words('english')
stop_words.extend(['rt', 'january', 'january6thcommitteehearings', 'ja', 'january6thcomm'])
tweets['cleaned'] = tweets['cleaned'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop_words)]))

### Tokenize Words in Cleaned Column

In [4]:
tweets['tokenized'] = tweets.apply(lambda row: nltk.word_tokenize(row['cleaned']), axis=1)

In [5]:
tweets.head()

text  \
0  @ReallyAmerican1 #Roevember and\n#ForThePeople...   
1  RT @sandibachom: IS THIS THING ON???!!This is ...   
2  RT @sandibachom: IS THIS THING ON???!!This is ...   
3  RT @tleehumphrey: Today is the beginning of th...   
4  RT @AdamKinzinger: Mitch McConnell.\nKevin McC...   

                                             cleaned  \
0  roevember forthepeople votebluein2022 standwit...   
1  thing pathetic acting sec defense chris miller...   
2  thing pathetic acting sec defense chris miller...   
3  today beginning inquiry trudeau gov use emerge...   
4  mitch mcconnell kevin mccarthy knew trump resp...   

                                           tokenized  
0  [roevember, forthepeople, votebluein2022, stan...  
1  [thing, pathetic, acting, sec, defense, chris,...  
2  [thing, pathetic, acting, sec, defense, chris,...  
3  [today, beginning, inquiry, trudeau, gov, use,...  
4  [mitch, mcconnell, kevin, mccarthy, knew, trum...

In [6]:
corpus = list(tweets.cleaned)

In [7]:
# turn corpus into a document-term matrix
vectorizer = CountVectorizer()
doc_term = vectorizer.fit_transform(corpus)

## LSA + TF-IDF 

### TF-IDF Vectorizer
- Instead of counts filling up matrix each value will be a TF-IDF value which is the term frequency or the percent of the document that consists of a particular term times the inverse document frequency, which is how rare the term is. 

In [8]:
#get the doc term matrix populated with the TF-IDF scores for the text corpus.
tfidf_vect = TfidfVectorizer()
tf_doc_term = tfidf_vect.fit_transform(corpus).toarray()
tf_doc_term_df = pd.DataFrame(tf_doc_term, columns = tfidf_vect.get_feature_names_out())

In [9]:
tf_doc_term.shape

(34993, 9399)

In [10]:
pd.set_option('display.max_columns', None)

In [11]:
#tf_doc_term_df.iloc[:, :300]

### LSA: Reduce Dimensionality 
- LSA is just SVD on the document term matrix

In [12]:
#number of topics indicated in arg of TruncatedSVD()
lsa2 = TruncatedSVD(n_components = 4, random_state = 122)
lsa2.fit(tf_doc_term)
lsa2

TruncatedSVD(n_components=4, random_state=122)

In [13]:
components=lsa2.components_
vocab = tfidf_vect.get_feature_names_out()

In [14]:
tf_topic_term = lsa2.components_.round(3)
tf_topic_term

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

#### Turn topic_term into dataframe

In [15]:
tf_topic_term_df = pd.DataFrame(tf_topic_term.round(3),
                index = ["component_1", "component_2", "component_3", "component_4"],
                columns = tfidf_vect.get_feature_names_out())
tf_topic_term_df

00  000   01   05   08   09   10    100  1000  1000x  100s  \
component_1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.000   0.0    0.0   0.0   
component_2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.001   0.0    0.0   0.0   
component_3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.001  -0.0    0.0   0.0   
component_4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.001   0.0    0.0   0.0   

             100yds  101  102  10am  10h  10percenttothebigguy  10pm  10th  \
component_1     0.0  0.0  0.0   0.0  0.0                   0.0   0.0   0.0   
component_2     0.0  0.0  0.0   0.0 -0.0                   0.0   0.0   0.0   
component_3     0.0 -0.0 -0.0   0.0 -0.0                   0.0   0.0   0.0   
component_4     0.0  0.0 -0.0   0.0 -0.0                   0.0   0.0   0.0   

                11  118  1199  11thhour   12  12deoctubre  12h   13  139  \
component_1  0.002  0.0   0.0       0.0  0.0          0.0  0.0  0.0  0.0   
component_2  0.009  0.0   0.0       0.0  0.0          0.0  0.0  0.0  0.0   
component_3  0.013  0.0   0.0       0.0  0.0          0.0  0.0 -0.0  0.0   
component_4  0.040  0.0   0.0       0.0  0.0          0.0  0.0  0.0  0.0   

             13th   14    147  14a  14th  14x   15  150  154  157  15s   16  \
component_1   0.0  0.0  0.000  0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
component_2   0.0  0.0  0.010  0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
component_3   0.0  0.0 -0.003  0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
component_4   0.0  0.0  0.012  0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

              17  1776   18  180  182  1857  187  187minutes   19  1961  1973  \
component_1  0.0   0.0  0.0  0.0  0.0   0.0  0.0         0.0  0.0   0.0   0.0   
component_2  0.0   0.0  0.0  0.0  0.0   0.0  0.0         0.0  0.0   0.0   0.0   
component_3 -0.0   0.0  0.0  0.0  0.0   0.0  0.0         0.0  0.0   0.0   0.0   
component_4 -0.0   0.0  0.0  0.0  0.0   0.0  0.0        -0.0  0.0   0.0   0.0   

             1980  1984   1a  1pm  1st   20  2007  2008  2011  2015  2016  \
component_1   0.0   0.0  0.0  0.0  0.0  0.0   0.0   0.0   0.0   0.0   0.0   
component_2   0.0   0.0  0.0  0.0  0.0  0.0   0.0   0.0   0.0   0.0   0.0   
component_3   0.0  -0.0  0.0  0.0 -0.0  0.0   0.0   0.0   0.0   0.0   0.0   
component_4   0.0  -0.0 -0.0  0.0  0.0  0.0   0.0  -0.0   0.0   0.0   0.0   

             2017  2018  2019  202   2020   2021   2022  2022midterms  2023  \
component_1   0.0   0.0   0.0  0.0  0.000  0.000  0.000           0.0   0.0   
component_2   0.0   0.0   0.0  0.0  0.001  0.000  0.000           0.0   0.0   
component_3   0.0   0.0   0.0  0.0  0.001  0.000  0.000           0.0   0.0   
component_4   0.0   0.0   0.0  0.0  0.002  0.001  0.001          -0.0   0.0   

             2024  2030  20th   21  215  216   22   23  23pm   24  240  246  \
component_1   0.0   0.0   0.0  0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0   
component_2   0.0   0.0   0.0  0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0   
component_3   0.0   0.0   0.0  0.0  0.0  0.0 -0.0  0.0   0.0  0.0 -0.0  0.0   
component_4   0.0   0.0   0.0  0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0   

              25  256  25th   26  261  271   28   29  292  29m   2a   2d  \
component_1  0.0  0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
component_2  0.0  0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
component_3  0.0  0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0 -0.0  0.0  0.0   
component_4  0.0  0.0   0.0 -0.0  0.0  0.0  0.0  0.0 -0.0  0.0  0.0  0.0   

             2day  2ndwin  2pro   30    30k   31  31st   32   33  350  360  \
component_1   0.0     0.0   0.0  0.0  0.000  0.0   0.0  0.0  0.0  0.0  0.0   
component_2   0.0     0.0   0.0  0.0  0.002  0.0   0.0  0.0  0.0  0.0  0.0   
component_3   0.0     0.0   0.0  0.0  0.002  0.0   0.0 -0.0  0.0  0.0 -0.0   
component_4   0.0     0.0  -0.0  0.0  0.004  0.0   0.0  0.0  0.0  0.0 -0.0   

             37thcircuitcourt   38  3am   3d  3rd   3x     40  401k  4092  \
component_1               0.0  0

#### Easier to read: Function displays the top terms in each topic 

In [16]:
def display_topics(model, feature_names, no_top_words, topic_names = None): 
    for ix, topic in enumerate(model.components_):
        if not topic_names or not topic_names[ix]:
            print("\nTopic ", ix + 1)
        else:
            print("\nTopic: ", topic_names[ix])
        print(", ".join([feature_names[i]
                        for i in topic.argsort()[:-no_top_words - 1:-1]]))
    print("\n")
    return model, feature_names, no_top_words

#### Get 8 terms for each topic

In [17]:
# number is the number of top terms you want to see
output = display_topics(lsa2, tfidf_vect.get_feature_names_out(), 8)
output


Topic  1
kevin, backed, mccarthy, mitch, mcconnell, responsible, called, knew

Topic  2
mark, democracy, violence, creating, hamill, correct, overthrowing, love

Topic  3
decided, rudy, hatch, giuliani, meadows, lost, yet, stone

Topic  4
demands, oath, unanimously, deserves, history, voted, subpoena, testify




(TruncatedSVD(n_components=4, random_state=122),
 array(['00', '000', '01', ..., 'zoom', 'zuma', 'zzy'], dtype=object),
 8)

#### Name the topics

In [18]:
display_topics(lsa2, tfidf_vect.get_feature_names_out(), 8, ['Mitch_Kevin_Knew', 'Hamill_Democracy','Meadows_Guiliani_Stone','Subpoena_Trump']);


Topic:  Mitch_Kevin_Knew
kevin, backed, mccarthy, mitch, mcconnell, responsible, called, knew

Topic:  Hamill_Democracy
mark, democracy, violence, creating, hamill, correct, overthrowing, love

Topic:  Meadows_Guiliani_Stone
decided, rudy, hatch, giuliani, meadows, lost, yet, stone

Topic:  Subpoena_Trump
demands, oath, unanimously, deserves, history, voted, subpoena, testify




#### Determine which topics are in each document. Transform the original doc_term matrix into a document-topic matrix

In [19]:
doc_topic2 = lsa2.transform(doc_term)
doc_topic2.shape

(34993, 4)

#### Turn doc_topic matrix into a dataframe

In [20]:
doc_topic_df2 = pd.DataFrame(doc_topic2.round(5), index = corpus, columns = ["Mitch_Kevin_Knew", "Hamill_Democracy","Meadows_Guiliani_Stone","Subpoena_Trump"])

In [21]:
doc_topic_df2.reset_index(inplace=True)
doc_topic_df2 = doc_topic_df2.rename(columns = {'index':'Processed_Tweets'})

In [22]:
doc_topic_df2.head()

Processed_Tweets  Mitch_Kevin_Knew  \
0  roevember forthepeople votebluein2022 standwit...           0.00002   
1  thing pathetic acting sec defense chris miller...           0.00020   
2  thing pathetic acting sec defense chris miller...           0.00020   
3  today beginning inquiry trudeau gov use emerge...           0.00026   
4  mitch mcconnell kevin mccarthy knew trump resp...           2.98070   

   Hamill_Democracy  Meadows_Guiliani_Stone  Subpoena_Trump  
0           0.00016                 0.00011         0.00036  
1           0.00288                 0.00909         0.00835  
2           0.00288                 0.00909         0.00835  
3           0.00242                 0.00319         0.00389  
4           0.01766                 0.03322         0.05555

In [23]:
# add original tweets to df
doc_topic_df2['Tweets'] = tweets['text']

In [24]:
#list(doc_topic_df2)

In [25]:
#change order of columns
doc_topic_df2 = doc_topic_df2.reindex(['Tweets','Processed_Tweets','Mitch_Kevin_Knew','Hamill_Democracy', 'Meadows_Guiliani_Stone', 'Subpoena_Trump' ], axis=1)

In [26]:
doc_topic_df2.head()

Tweets  \
0  @ReallyAmerican1 #Roevember and\n#ForThePeople...   
1  RT @sandibachom: IS THIS THING ON???!!This is ...   
2  RT @sandibachom: IS THIS THING ON???!!This is ...   
3  RT @tleehumphrey: Today is the beginning of th...   
4  RT @AdamKinzinger: Mitch McConnell.\nKevin McC...   

                                    Processed_Tweets  Mitch_Kevin_Knew  \
0  roevember forthepeople votebluein2022 standwit...           0.00002   
1  thing pathetic acting sec defense chris miller...           0.00020   
2  thing pathetic acting sec defense chris miller...           0.00020   
3  today beginning inquiry trudeau gov use emerge...           0.00026   
4  mitch mcconnell kevin mccarthy knew trump resp...           2.98070   

   Hamill_Democracy  Meadows_Guiliani_Stone  Subpoena_Trump  
0           0.00016                 0.00011         0.00036  
1           0.00288                 0.00909         0.00835  
2           0.00288                 0.00909         0.00835  
3           0.00242                 0.00319         0.00389  
4           0.01766                 0.03322         0.05555

#### Make new dataframe that filters values of topic columns
We're removing content with lowest scores
- Filters => 25% of total column value
- Filtering went from 34,993 tweets to 19,741

In [27]:
top_score_tweets = doc_topic_df2.copy()

In [28]:
filt1 = top_score_tweets['Mitch_Kevin_Knew'].quantile(0.25)
filt2 = top_score_tweets['Hamill_Democracy'].quantile(0.25)
filt3 = top_score_tweets['Meadows_Guiliani_Stone'].quantile(0.25)
filt4 = top_score_tweets['Subpoena_Trump'].quantile(0.25)

In [29]:
top_score_tweets = top_score_tweets[top_score_tweets['Mitch_Kevin_Knew'] >= filt1]
top_score_tweets = top_score_tweets[top_score_tweets['Hamill_Democracy'] >= filt2]
top_score_tweets = top_score_tweets[top_score_tweets['Meadows_Guiliani_Stone'] >= filt3]
top_score_tweets = top_score_tweets[top_score_tweets['Subpoena_Trump'] >= filt4]

In [30]:
top_score_tweets.head()

Tweets  \
4   RT @AdamKinzinger: Mitch McConnell.\nKevin McC...   
8   RT @2RawTooReal: I still cant get over the fac...   
11  RT @Navajo_6: You can’t tell me you have an ou...   
15  RT @StandForBetter: The Secret Service clearly...   
19  RT @Navajo_6: You can’t tell me you have an ou...   

                                     Processed_Tweets  Mitch_Kevin_Knew  \
4   mitch mcconnell kevin mccarthy knew trump resp...           2.98070   
8   still cant get fact nancy pelosi sh ts wanted ...           0.24355   
11  6 tell ounce brains head still believe trump p...           0.23625   
15  secret service clearly involved coverup trump ...           0.23554   
19  6 tell ounce brains head still believe trump p...           0.23625   

    Hamill_Democracy  Meadows_Guiliani_Stone  Subpoena_Trump  
4            0.01766                 0.03322         0.05555  
8            0.24108                 0.25057         0.56588  
11           0.16989                 0.20159         0.27820  
15           0.17287                 0.21018         0.27962  
19           0.16989                 0.20159         0.27820

In [31]:
max_values2 = top_score_tweets[['Mitch_Kevin_Knew','Hamill_Democracy','Meadows_Guiliani_Stone', 'Subpoena_Trump']].max(axis=1)

In [32]:
max_values2

4        2.98070
8        0.56588
11       0.27820
15       0.27962
19       0.27820
          ...   
34985    1.66830
34986    0.09197
34988    0.10602
34991    0.53460
34992    0.29832
Length: 19741, dtype: float64

In [33]:
#max_values = top_score_tweets.max(axis=1)

In [34]:
#max_values

In [35]:
# Select only the numeric columns from the DataFrame
numeric_columns = top_score_tweets.select_dtypes(include=[int, float])

In [36]:
# Find the largest value in each row of the numeric columns
largest_values_per_row = numeric_columns.max(axis=1)

In [37]:
# Create a boolean mask where True indicates the largest value in each row
is_largest_numeric = numeric_columns.eq(largest_values_per_row, axis=0).idxmax(axis=1)

In [38]:
is_largest_numeric

4        Mitch_Kevin_Knew
8          Subpoena_Trump
11         Subpoena_Trump
15         Subpoena_Trump
19         Subpoena_Trump
               ...       
34985      Subpoena_Trump
34986      Subpoena_Trump
34988      Subpoena_Trump
34991      Subpoena_Trump
34992      Subpoena_Trump
Length: 19741, dtype: object

#### Adds the 'is_largest_numeric' series to the dataframe

In [39]:
top_score_tweets['topic_label']=pd.Series(is_largest_numeric)

In [40]:
top_score_tweets.head()

Tweets  \
4   RT @AdamKinzinger: Mitch McConnell.\nKevin McC...   
8   RT @2RawTooReal: I still cant get over the fac...   
11  RT @Navajo_6: You can’t tell me you have an ou...   
15  RT @StandForBetter: The Secret Service clearly...   
19  RT @Navajo_6: You can’t tell me you have an ou...   

                                     Processed_Tweets  Mitch_Kevin_Knew  \
4   mitch mcconnell kevin mccarthy knew trump resp...           2.98070   
8   still cant get fact nancy pelosi sh ts wanted ...           0.24355   
11  6 tell ounce brains head still believe trump p...           0.23625   
15  secret service clearly involved coverup trump ...           0.23554   
19  6 tell ounce brains head still believe trump p...           0.23625   

    Hamill_Democracy  Meadows_Guiliani_Stone  Subpoena_Trump       topic_label  
4            0.01766                 0.03322         0.05555  Mitch_Kevin_Knew  
8            0.24108                 0.25057         0.56588    Subpoena_Trump  
11           0.16989                 0.20159         0.27820    Subpoena_Trump  
15           0.17287                 0.21018         0.27962    Subpoena_Trump  
19           0.16989                 0.20159         0.27820    Subpoena_Trump

In [41]:
top_score_tweets.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19741 entries, 4 to 34992
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Tweets                  19741 non-null  object 
 1   Processed_Tweets        19741 non-null  object 
 2   Mitch_Kevin_Knew        19741 non-null  float64
 3   Hamill_Democracy        19741 non-null  float64
 4   Meadows_Guiliani_Stone  19741 non-null  float64
 5   Subpoena_Trump          19741 non-null  float64
 6   topic_label             19741 non-null  object 
dtypes: float64(4), object(3)
memory usage: 1.2+ MB


In [42]:
simple_df = top_score_tweets.copy()
simple_df.drop(columns=['Mitch_Kevin_Knew', 'Hamill_Democracy', 'Meadows_Guiliani_Stone', 'Subpoena_Trump'], inplace=True)

In [43]:
simple_df.head()

Tweets  \
4   RT @AdamKinzinger: Mitch McConnell.\nKevin McC...   
8   RT @2RawTooReal: I still cant get over the fac...   
11  RT @Navajo_6: You can’t tell me you have an ou...   
15  RT @StandForBetter: The Secret Service clearly...   
19  RT @Navajo_6: You can’t tell me you have an ou...   

                                     Processed_Tweets       topic_label  
4   mitch mcconnell kevin mccarthy knew trump resp...  Mitch_Kevin_Knew  
8   still cant get fact nancy pelosi sh ts wanted ...    Subpoena_Trump  
11  6 tell ounce brains head still believe trump p...    Subpoena_Trump  
15  secret service clearly involved coverup trump ...    Subpoena_Trump  
19  6 tell ounce brains head still believe trump p...    Subpoena_Trump

In [44]:
simple_df.shape

(19741, 3)

In [45]:
remove_dupes = simple_df.copy()

In [46]:
remove_dupes.drop_duplicates(subset='Tweets', inplace=True)

In [47]:
remove_dupes.drop_duplicates(subset='Processed_Tweets', inplace=True)

In [48]:
remove_dupes.shape

(2043, 3)

In [49]:
remove_dupes.head(100)

Tweets  \
4    RT @AdamKinzinger: Mitch McConnell.\nKevin McC...   
8    RT @2RawTooReal: I still cant get over the fac...   
11   RT @Navajo_6: You can’t tell me you have an ou...   
15   RT @StandForBetter: The Secret Service clearly...   
24   RT @ifindkarma: “Trump’s claim that Pelosi fai...   
..                                                 ...   
732  @BradMossEsq It wasn’t even willful blindness....   
737  RT @ifindkarma: “[Trump reacting to the Suprem...   
745  And here we have House Speaker Nancy Pelosi sh...   
749  RT @GoodTroubleShow: We are Livestreaming the ...   
840  Donald Trump knew he lost.” 2/2 - Chmn. Bennie...   

                                      Processed_Tweets             topic_label  
4    mitch mcconnell kevin mccarthy knew trump resp...        Mitch_Kevin_Knew  
8    still cant get fact nancy pelosi sh ts wanted ...          Subpoena_Trump  
11   6 tell ounce brains head still believe trump p...          Subpoena_Trump  
15   secret service clearly involved coverup trump ...          Subpoena_Trump  
24   trump claim pelosi failed call law enforcement...  Meadows_Guiliani_Stone  
..                                                 ...                     ...  
732  even willful blindness since 9 11 law enforcem...          Subpoena_Trump  
737  trump reacting supreme court rejecting electio...          Subpoena_Trump  
745  house speaker nancy pelosi showing concern vp ...          Subpoena_Trump  
749  livestreaming 6th hearing youtube 1pm eastern ...          Subpoena_Trump  
840  donald trump knew lost 2 2 chmn bennie thompso...        Mitch_Kevin_Knew  

[100 rows x 3 columns]

## Batch Work New Flow
- To make this work the improvements: 
    - pull out topic as a keyword because you want check intrusions for each topic. Instead of random choice make it an arg. 
- video 1:01:05 improve the code by mixing the order of the dictionaries in the batch so not so easy to predict. Currently intruder always last
- Input: Pandas Dataframe
- Output: List of Dictionaries
- -> List[Dict[str, str]]
- Expected_result = [{“document”: doc1, “topic”: topic 1, “intruder”: 0}, {“document”: doc2, “topic”: topic 1, “intruder”: 0}, {“document”: doc3, “topic”: topic 1, “intruder”: 0}, {“document”: doc4, “topic”: topic 1, “intruder”: 0}, {“document”: doc120, “topic”: NOT topic 1, “intruder”: 1}]

In [123]:
def create_batch(df, main_topic, specific_topic):
    # Get unique topics
    topics = df['topic_label'].unique()

    # Check if the main_topic is in the topics
    if main_topic not in topics:
        raise ValueError(f"Main topic '{main_topic}' not found in the DataFrame.")
    
    batch_topic_df = df[df['topic_label'] == main_topic]
    batch = batch_topic_df.sample(n=4).to_dict('records')
    
    # Mark these documents as not intruders
    for doc in batch:
        doc['intruder'] = 0

    # Select intruder documents from another topic
    intruder_df = df[df['topic_label'] == specific_topic].sample(1)

    # Converting ths sampled 'intruder_df' into a list of dicts and marking as intruders
    intruders = intruder_df.to_dict('records')  
    for intruder in intruders:
        intruder['intruder'] = 1

    # Add the intruder document to the batch
    batch.extend(intruders)
    
    # Shuffle the batch to randomly position the intruder
    random.shuffle(batch)

    return batch


### Run the function on different topics

In [124]:
#topics
remove_dupes.topic_label.unique()

array(['Mitch_Kevin_Knew', 'Subpoena_Trump', 'Meadows_Guiliani_Stone',
       'Hamill_Democracy'], dtype=object)

#### Batch 1 Mitch Kevin Knew / Intruder 1

In [125]:
#creates list of dicts
mitch_kev_1 = create_batch(remove_dupes, 'Mitch_Kevin_Knew', 'Hamill_Democracy')
#creates dataframe from list of dicts
mitch_kev_1 = pd.DataFrame.from_records(mitch_kev_1)

In [126]:
mitch_kev_1

Tweets  \
0  @The_FJC That #January6thCommitteeHearings was...   
1  @washingtonpost #America #AmericaFirst #Januar...   
2  Donald Trump knew he lost.” 2/2 - Chmn. Bennie...   
3  These #January6thCommitteeHearings have been e...   
4  RT @ColorMeStevie: Mr. Trump is responsible fo...   

                                    Processed_Tweets       topic_label  \
0  fjc planned yesterday exactly reason knew repo...  Mitch_Kevin_Knew   
1      america americafirst making martyr trump maga  Hamill_Democracy   
2  donald trump knew lost 2 2 chmn bennie thompso...  Mitch_Kevin_Knew   
3  explosive knew jan 6 premeditated tfg amp team...  Mitch_Kevin_Knew   
4           mr trump responsible 6th period end lock  Mitch_Kevin_Knew   

   intruder  
0         0  
1         1  
2         0  
3         0  
4         0

#### Batch 1 Mitch Kevin Knew / Intruder 2

In [127]:
#creates list of dicts
mitch_kev_2 = create_batch(remove_dupes, 'Mitch_Kevin_Knew', 'Subpoena_Trump')
mitch_kev_2 = pd.DataFrame.from_records(mitch_kev_2)

In [128]:
mitch_kev_2

Tweets  \
0  And here is weasel Mitch McConnell …in April 2...   
1  On Jan 6 speaker Pelosi was President pelosi s...   
2  RT @taradublinrocks: Trump knew he lost. \nTru...   
3  The committee presented all the evidence. Trum...   
4  FBI Director Chris Wray IS PART of the #Republ...   

                                    Processed_Tweets       topic_label  \
0  weasel mitch mcconnell april 2022 said would s...  Mitch_Kevin_Knew   
1  jan 6 speaker pelosi president pelosi one trul...    Subpoena_Trump   
2  trump knew lost trump team knew lost kept goin...  Mitch_Kevin_Knew   
3  committee presented evidence trump lap dogs st...  Mitch_Kevin_Knew   
4  fbi director chris wray part republicans coup ...  Mitch_Kevin_Knew   

   intruder  
0         0  
1         1  
2         0  
3         0  
4         0

#### Batch 1 Mitch Kevin Knew / Intruder 3

In [129]:
mitch_kev_3 = create_batch(remove_dupes, 'Mitch_Kevin_Knew', 'Meadows_Guiliani_Stone')
mitch_kev_3 = pd.DataFrame.from_records(mitch_kev_3)

In [130]:
mitch_kev_3

Tweets  \
0  @djrothkopf Why is this a surprise when we alr...   
1  Trump is responsible for the insurrection and ...   
2  RT @dudeonthebay: Trump to McCarthy - “They’re...   
3  “DJT knew he lost.  And he did everything he c...   
4  RT @argus_constant: @Justme10000000 @waltshaub...   

                                    Processed_Tweets             topic_label  \
0  surprise already knew white supremacists domes...        Mitch_Kevin_Knew   
1  trump responsible insurrection finally must te...        Mitch_Kevin_Knew   
2  trump mccarthy people antifa mccarthy sir peop...        Mitch_Kevin_Knew   
3  djt knew lost everything could prevent certifi...        Mitch_Kevin_Knew   
4  constant trump defender excuses knowing lost d...  Meadows_Guiliani_Stone   

   intruder  
0         0  
1         0  
2         0  
3         0  
4         1

#### Batch 2 Hamill_Democracy / Intruder 1

In [131]:
hamill_1 = create_batch(remove_dupes, 'Hamill_Democracy', 'Mitch_Kevin_Knew')
hamill_1 = pd.DataFrame.from_records(hamill_1)

In [132]:
hamill_1

Tweets  \
0  @robert_burnier And Biden managed to do all th...   
1  Chilling video with the #January6thCommitteeHe...   
2  “DJT knew he lost.  And he did everything he c...   
3  RT @GerberKawasaki: This is insane footage of ...   
4  @SpeakerPelosi was calm and getting organized ...   

                                    Processed_Tweets       topic_label  \
0  burnier biden managed fraud trump charge count...  Hamill_Democracy   
1  chilling video today proves joshhawley cut amp...  Hamill_Democracy   
2  djt knew lost everything could prevent certifi...  Mitch_Kevin_Knew   
3  insane footage attack democracy desperation he...  Hamill_Democracy   
4  calm getting organized like queen control mean...  Hamill_Democracy   

   intruder  
0         0  
1         0  
2         1  
3         0  
4         0

#### Batch 2 Hamill_Democracy / Intruder 2

In [133]:
hamill_2 = create_batch(remove_dupes, 'Hamill_Democracy', 'Subpoena_Trump')
hamill_2 = pd.DataFrame.from_records(hamill_2)

In [134]:
hamill_2

Tweets  \
0  Watching the #January6thCommitteeHearings and ...   
1  Its hilarious that Trump is being subpoenaed f...   
2  When the violent armed insurrectionists were c...   
3  #January6thCommitteeHearings This is why Jaime...   
4  I woke up this morning in cold sweats just rem...   

                                    Processed_Tweets       topic_label  \
0  watching seeing taking action restore order co...  Hamill_Democracy   
1  hilarious trump subpoenaed jan6 hearing needs ...  Hamill_Democracy   
2  violent armed insurrectionists coming stepped ...  Hamill_Democracy   
3  jaime lost primary right thing needs win maga ...  Hamill_Democracy   
4  woke morning cold sweats remembering happened ...    Subpoena_Trump   

   intruder  
0         0  
1         0  
2         0  
3         0  
4         1

#### Batch 2 Hamill_Democracy / Intruder 3

In [135]:
hamill_3 = create_batch(remove_dupes, 'Hamill_Democracy', 'Meadows_Guiliani_Stone')
hamill_3 = pd.DataFrame.from_records(hamill_3)

In [136]:
hamill_3

Tweets  \
0  You've been conned, #MAGA.\nYou continue to be...   
1  @katiehobbs Trump admitted he lied and knows h...   
2  @TheLeoTerrell Only moral cowards would run te...   
3  RT @jossytweets: I bet Trump is gonna start a ...   
4  The future is predictable. \nBy the way, Trump...   

                                    Processed_Tweets             topic_label  \
0  conned maga continue conned trump knew lost li...        Hamill_Democracy   
1  trump admitted lied knows lost lake whole camp...        Hamill_Democracy   
2  moral cowards would run terrified watching tv ...        Hamill_Democracy   
3  bet trump gonna start new conspiracy democrats...  Meadows_Guiliani_Stone   
4  future predictable way trump coming trump puti...        Hamill_Democracy   

   intruder  
0         0  
1         0  
2         0  
3         1  
4         0

#### Batch 3 Meadows_Guiliani_Stone / Intruder 1

In [137]:
mgs_1 = create_batch(remove_dupes, 'Meadows_Guiliani_Stone', 'Mitch_Kevin_Knew')
mgs_1 = pd.DataFrame.from_records(mgs_1)

In [138]:
mgs_1

Tweets  \
0  RT @DeanObeidallah: In no other first world na...   
1  Nancy Pelosi is BadAss 81 at the time not Scar...   
2  RT @joekatz45: Trump: "I literally don't care ...   
3  RT @Jemappelleamie: Dear Donald, you lost Mich...   
4  @FoxNews The Big Takeaway from today’s hearing...   

                                    Processed_Tweets             topic_label  \
0  first world nation would donald trump able att...  Meadows_Guiliani_Stone   
1  nancy pelosi badass 81 time scared corner hidi...  Meadows_Guiliani_Stone   
2  trump literally care live die kevin mccarthy f...        Mitch_Kevin_Knew   
3      dear donald lost michigan pos signed michigan  Meadows_Guiliani_Stone   
4  big takeaway today hearing trump attempted cou...  Meadows_Guiliani_Stone   

   intruder  
0         0  
1         0  
2         1  
3         0  
4         0

#### Batch 3 Meadows_Guiliani_Stone / Intruder 2

In [139]:
mgs_2 = create_batch(remove_dupes, 'Meadows_Guiliani_Stone', 'Subpoena_Trump')
mgs_2 = pd.DataFrame.from_records(mgs_2)

In [140]:
mgs_2

Tweets  \
0  COMMUNISTA BLUMENTHAL FALSE CLAIM CHARGES &amp...   
1  RT @mayawiley: This footage. Pelosi, Schumer t...   
2  Anyone else stunned to learn Pelosi, Schumer &...   
3  It is time for Democrats to get tough on crime...   
4  “What kind of snowflake is Donald Trump that h...   

                                    Processed_Tweets             topic_label  \
0  communista blumenthal false claim charges amp ...  Meadows_Guiliani_Stone   
1  footage pelosi schumer trying get help capitol...  Meadows_Guiliani_Stone   
2  anyone else stunned learn pelosi schumer amp s...  Meadows_Guiliani_Stone   
3  time democrats get tough crime people need get...  Meadows_Guiliani_Stone   
4  kind snowflake donald trump willing send peopl...          Subpoena_Trump   

   intruder  
0         0  
1         0  
2         0  
3         0  
4         1

#### Batch 3 Meadows_Guiliani_Stone / Intruder 3

In [141]:
mgs_3 = create_batch(remove_dupes, 'Meadows_Guiliani_Stone', 'Hamill_Democracy')
mgs_3 = pd.DataFrame.from_records(mgs_3)

In [142]:
mgs_3

Tweets  \
0  RT @BrennanCenter: Rep. Schiff: "Evidence show...   
1  Nancy Pelosi is BadAss 81 at the time not Scar...   
2  @TomFitton 🧵NEW from me: Tom Fitton, the far-r...   
3  @SpiroAgnewGhost When you watch that video, sh...   
4  COMMUNISTA BLUMENTHAL VIETNAM SERVICE LIE &amp...   

                                    Processed_Tweets             topic_label  \
0  rep schiff evidence shows trump aware risk vio...        Hamill_Democracy   
1  nancy pelosi badass 81 time scared corner hidi...  Meadows_Guiliani_Stone   
2  new tom fitton far right fake lawyer got trump...  Meadows_Guiliani_Stone   
3  watch video look crazy watch pretty clear craz...  Meadows_Guiliani_Stone   
4  communista blumenthal vietnam service lie amp ...  Meadows_Guiliani_Stone   

   intruder  
0         1  
1         0  
2         0  
3         0  
4         0

#### Batch 4 Subpoena Trump / Intruder 1

In [143]:
sub_tr_1 = create_batch(remove_dupes, 'Subpoena_Trump', 'Mitch_Kevin_Knew')
sub_tr_1 = pd.DataFrame.from_records(sub_tr_1)

In [144]:
sub_tr_1

Tweets  \
0  RT @jbf1755: Wow -- supposedly, the Committee ...   
1  On the same day treasonous Donald Trump was su...   
2  https://t.co/SYktQBD6en\n#Jan6thInsurrection  ...   
3  Since #Trump was subpoena there should be a 10...   
4  @MacFarlaneNews I tried but I cannot read this...   

                                    Processed_Tweets       topic_label  \
0  wow supposedly committee vote trump subpoena h...    Subpoena_Trump   
1  day treasonous donald trump subpoenaed supreme...    Subpoena_Trump   
2  jan6thinsurrection 6 committee votes subpoena ...    Subpoena_Trump   
3  since trump subpoena 10th primetime world coul...    Subpoena_Trump   
4  tried cannot read nonsense revealed need know ...  Mitch_Kevin_Knew   

   intruder  
0         0  
1         0  
2         0  
3         0  
4         1

In [145]:
sub_tr_2 = create_batch(remove_dupes, 'Subpoena_Trump', 'Meadows_Guiliani_Stone')
sub_tr_2 = pd.DataFrame.from_records(sub_tr_2)

In [146]:
sub_tr_2

Tweets  \
0  RT @MmeDefargeKnits: Good #Catholic girl Nancy...   
1  https://t.co/VqPQuVsbx6\nDINOBROS President Do...   
2  RT @DrDinD: As the #January6thCommitteeHearing...   
3  RT @petenet321: "President Donald Trump"\n#Jan...   
4  RT @d08ef2df608d4a8: So Trump tried to sabotag...   

                                    Processed_Tweets             topic_label  \
0  good catholic girl nancy pelosi say word shit ...  Meadows_Guiliani_Stone   
1  dinobros president donald trump 2020 toy figur...          Subpoena_Trump   
2  continue today reminder cannot election denier...          Subpoena_Trump   
3  president donald trump president trump loves i...          Subpoena_Trump   
4  trump tried sabotage us troops immediate withd...          Subpoena_Trump   

   intruder  
0         1  
1         0  
2         0  
3         0  
4         0

In [147]:
sub_tr_3 = create_batch(remove_dupes, 'Subpoena_Trump', 'Hamill_Democracy')
sub_tr_3 = pd.DataFrame.from_records(sub_tr_3)

In [148]:
sub_tr_3

Tweets  \
0  @SpeakerPelosi is doing what Trump should have...   
1  @DonaldJTrumpJr please pass a message on to yo...   
2  #January6thCommitteeHearings are such a stage ...   
3  I'd rather see baby formula on the shelves tha...   
4  #January6thCommitteeHearings This is why Jaime...   

                                    Processed_Tweets       topic_label  \
0  trump charge jan6thinsurrection jan6thhearings...    Subpoena_Trump   
1  please pass message dad lest see dad balls tes...    Subpoena_Trump   
2  stage performance staged pelosi calls released...    Subpoena_Trump   
3   rather see baby formula shelves trump testifying    Subpoena_Trump   
4  jaime lost primary right thing needs win maga ...  Hamill_Democracy   

   intruder  
0         0  
1         0  
2         0  
3         0  
4         1

#### All Batches Into Dataframe

In [149]:
frames = [mitch_kev_1,
mitch_kev_2,
mitch_kev_3,
hamill_1,
hamill_2,
hamill_3,
mgs_1,
mgs_2,
mgs_3,
sub_tr_1,
sub_tr_2,
sub_tr_3]

In [150]:
result = pd.concat(frames)

In [151]:
# reset index
result.reset_index(drop=True, inplace=True)

In [152]:
result.head()

Tweets  \
0  @The_FJC That #January6thCommitteeHearings was...   
1  @washingtonpost #America #AmericaFirst #Januar...   
2  Donald Trump knew he lost.” 2/2 - Chmn. Bennie...   
3  These #January6thCommitteeHearings have been e...   
4  RT @ColorMeStevie: Mr. Trump is responsible fo...   

                                    Processed_Tweets       topic_label  \
0  fjc planned yesterday exactly reason knew repo...  Mitch_Kevin_Knew   
1      america americafirst making martyr trump maga  Hamill_Democracy   
2  donald trump knew lost 2 2 chmn bennie thompso...  Mitch_Kevin_Knew   
3  explosive knew jan 6 premeditated tfg amp team...  Mitch_Kevin_Knew   
4           mr trump responsible 6th period end lock  Mitch_Kevin_Knew   

   intruder  
0         0  
1         1  
2         0  
3         0  
4         0

In [153]:
#create column with intruder index

In [154]:
result['intruder_index'] = None

In [155]:
# Assign the index to the new column where column 'B' is True
result.loc[result['intruder'] == True, 'intruder_index'] = result.index[result['intruder'] == True]

In [156]:
result

Tweets  \
0   @The_FJC That #January6thCommitteeHearings was...   
1   @washingtonpost #America #AmericaFirst #Januar...   
2   Donald Trump knew he lost.” 2/2 - Chmn. Bennie...   
3   These #January6thCommitteeHearings have been e...   
4   RT @ColorMeStevie: Mr. Trump is responsible fo...   
5   And here is weasel Mitch McConnell …in April 2...   
6   On Jan 6 speaker Pelosi was President pelosi s...   
7   RT @taradublinrocks: Trump knew he lost. \nTru...   
8   The committee presented all the evidence. Trum...   
9   FBI Director Chris Wray IS PART of the #Republ...   
10  @djrothkopf Why is this a surprise when we alr...   
11  Trump is responsible for the insurrection and ...   
12  RT @dudeonthebay: Trump to McCarthy - “They’re...   
13  “DJT knew he lost.  And he did everything he c...   
14  RT @argus_constant: @Justme10000000 @waltshaub...   
15  @robert_burnier And Biden managed to do all th...   
16  Chilling video with the #January6thCommitteeHe...   
17  “DJT knew he lost.  And he did everything he c...   
18  RT @GerberKawasaki: This is insane footage of ...   
19  @SpeakerPelosi was calm and getting organized ...   
20  Watching the #January6thCommitteeHearings and ...   
21  Its hilarious that Trump is being subpoenaed f...   
22  When the violent armed insurrectionists were c...   
23  #January6thCommitteeHearings This is why Jaime...   
24  I woke up this morning in cold sweats just rem...   
25  You've been conned, #MAGA.\nYou continue to be...   
26  @katiehobbs Trump admitted he lied and knows h...   
27  @TheLeoTerrell Only moral cowards would run te...   
28  RT @jossytweets: I bet Trump is gonna start a ...   
29  The future is predictable. \nBy the way, Trump...   
30  RT @DeanObeidallah: In no other first world na...   
31  Nancy Pelosi is BadAss 81 at the time not Scar...   
32  RT @joekatz45: Trump: "I literally don't care ...   
33  RT @Jemappelleamie: Dear Donald, you lost Mich...   
34  @FoxNews The Big Takeaway from today’s hearing...   
35  COMMUNISTA BLUMENTHAL FALSE CLAIM CHARGES &amp...   
36  RT @mayawiley: This footage. Pelosi, Schumer t...   
37  Anyone else stunned to learn Pelosi, Schumer &...   
38  It is time for Democrats to get tough on crime...   
39  “What kind of snowflake is Donald Trump that h...   
40  RT @BrennanCenter: Rep. Schiff: "Evidence show...   
41  Nancy Pelosi is BadAss 81 at the time not Scar...   
42  @TomFitton 🧵NEW from me: Tom Fitton, the far-r...   
43  @SpiroAgnewGhost When you watch that video, sh...   
44  COMMUNISTA BLUMENTHAL VIETNAM SERVICE LIE &amp...   
45  RT @jbf1755: Wow -- supposedly, the Committee ...   
46  On the same day treasonous Donald Trump was su...   
47  https://t.co/SYktQBD6en\n#Jan6thInsurrection  ...   
48  Since #Trump was subpoena there should be a 10...   
49  @MacFarlaneNews I tried but I cannot read this...   
50  RT @MmeDefargeKnits: Good #Catholic girl Nancy...   
51  https://t.co/VqPQuVsbx6\nDINOBROS President Do...   
52  RT @DrDinD: As the #January6thCommitteeHearing...   
53  RT @petenet321: "President Donald Trump"\n#Jan...   
54  RT @d08ef2df608d4a8: So Trump tried to sabotag...   
55  @SpeakerPelosi is doing what Trump should have...   
56  @DonaldJTrumpJr please pass a message on to yo...   
57  #January6thCommitteeHearings are such a stage ...   
58  I'd rather see baby formula on the shelves tha...   
59  #January6thCommitteeHearings This is why Jaime...   

                                     Processed_Tweets             topic_label  \
0   fjc planned yesterday exactly reason knew repo...        Mitch_Kevin_Knew   
1       america americafirst making martyr trump maga        Hamill_Democracy   
2   donald trump knew lost 2 2 chmn bennie thompso...        Mitch_Kevin_Knew   
3   explosive knew jan 6 premeditated tfg amp team...        Mitch_Kevin_Knew   
4            mr trump responsible 6th period end lock        Mitch_Kevin_Knew   
5   weasel mitch mcconnell april 2022 said would s...        Mitch_Kevin_Knew   
6 

#### CSVs

In [157]:
#result.to_csv('result_draft_final.csv')

In [85]:
#mitch_kev.to_csv('batch1_intruder1.csv', index=False)

### Test it's working as intended

The test will verify that the batch has exactly 5 documents and that only one of them is an intruder. Also, it will verify that the documents from the same topic are not intruders.
- Is this test good? How to improve
- Give you an idea of seeing how this is working

In [86]:
# import unittest
# from unittest.mock import patch
# import pytest
# import pandas as pd
# import random

# # Assuming create_batch() is in a file named my_module.py
# from my_module import create_batch

# class TestCreateBatch(unittest.TestCase):
#     @patch("random.shuffle")
#     def test_create_batch(self, mock_shuffle):
#         # Mock shuffle function to do nothing (to make the test deterministic)
#         mock_shuffle.side_effect = lambda x: x

#         # Create a mock dataframe for testing
#         data = {
#             'topic_label': ['topic1', 'topic1', 'topic1', 'topic1', 'topic1', 'topic2'],
#             'content': ['doc1', 'doc2', 'doc3', 'doc4', 'doc5', 'intruder']
#         }
#         df = pd.DataFrame(data)

#         # Use the function
#         batch = create_batch(df, 'topic1', 'topic2')

#         # Test if the batch has exactly 5 documents
#         self.assertEqual(len(batch), 5)

#         # Test if only one of them is an intruder
#         intruder_count = sum(1 for doc in batch if doc['intruder'] == 1)
#         self.assertEqual(intruder_count, 1)

#         # Test if the documents from the same topic are not intruders
#         for doc in batch:
#             if doc['topic_label'] == 'topic1':
#                 self.assertEqual(doc['intruder'], 0)

# # Run the test
# if __name__ == '__main__':
#     unittest.main()


In [87]:
#You would put this code in a file, say test_create_batch.py, and run it with the command pytest test_create_batch.py. Please note that you need to have the pytest module installed. You can install it using pip:
# pip install pytest
# If the tests pass, you can be confident that your function is working correctly. If any of the tests fail, pytest will provide detailed output that can help you identify the issue.

# import pytest

# def test_create_batch():
#     # Setup
#     df = pd.read_csv('/mnt/data/simple_df.csv')

#     # Test that the function returns a list of length 5
#     batch = create_batch(df, 'Subpoena_Trump')
#     assert len(batch) == 5

#     # Test that the function returns 4 documents with "intruder" set to 0 and 1 document with "intruder" set to 1
#     intruders = [doc['intruder'] for doc in batch]
#     assert intruders.count(0) == 4
#     assert intruders.count(1) == 1

#     # Test that the function returns 4 documents with the same topic and 1 document with a different topic
#     topics = [doc['topic'] for doc in batch]
#     assert topics.count('Subpoena_Trump') == 4
#     assert 'Subpoena_Trump' not in topics[-1]

#     # Test that the function raises a ValueError when the provided main topic is not in the dataframe
#     with pytest.raises(ValueError):
#         create_batch(df, 'Nonexistent_Topic')